### ISSUE TITLE
Linter: Refactor to decorator

### ISSUE THREAD

**Issue by AbdealiJK** [coala-analyzer/coala#1928]
The idea that @Makman2 is working on at https://github.com/coala-analyzer/coala/pull/1814 is to refactor the Lint into a decorator. There was no nice issue for this yet, and Im making this for discussion. (I don't want to have huge conversations like the one @Makman2 posted on yesterday on gitter. It'd get lost.)

So, in the PR I suggested we could use MetaClasses instead of decorators.

**Comment by AbdealiJK** [comment]
@Makman2 wrote:

alright @AbdealiJK @sils1297
I'm coming to the conclusion that metaclasses are not the right way for the next-gen linter design
why?
After some hours trying to understand how metaclasses actually work, the principle is "the metaclass is the type of a class"
so when a class is created (**not instantiated!**) for example via a class declaration like this:
```
 class X:
    pass
```
the metaclass constructor is invoked that takes care of creating this class `X`.
By default `type` is the metaclass for every class, and already implements some sort of class generation of functionality (like turning functions into wrappers that are called together with `self` etc etc)
coming to why this doesn't work that well (at least it makes conceptually no sense):

My idea was to let the user define two main functions: `create_arguments` and `generate_config`
These get integrated into the linter workflow, means calling the underlying executable (via arguments from `create_arguments` and config from `generate_config`), parsing the stuff, yielding results
the class the user defines with `create_arguments` and `generate_config` was wrapped into an interface member of the actual linter class that does everything
e.g.:
```
class mylinter(Linter(executable, args, ...)): # <-- this is already the metaclass approach
    @staticmethod
    def create_arguments(file, filename, config_file):
        pass

mylinter._interface # <-- the actual class contents defined by the user
```
the metaclass would do this:
```
class LinterMetaclass(type):
    def __new__(cls, name, bases, dct):
        class LinterBaseFunctionality:
            _interface = type(name, bases, dct)

            def run():
                pass
            # Other stuff the base linter does

        # I hope this is okay instead of using `type.__new__`,  but it's just the principle
        return LinterBaseFunctionality
```

 - problem 1: This unnecessary self-using of type to construct the user-defined class
as we wrap the stuff only we still need to create it, doing this manually is just not nice
 - problem 2: Inheritance
and this is the actual main problem. Consider this:
```
class JavaModuleLinter(LinterMetaclass(executable="java", ...)):
    def create_arguments(file, filename, config_file):
        # Don't know whether this is a valid command, but it's just the principle :)
        return ("--module",)

class MyJavaLinter(JavaModuleLinter):
    def create_arguments(file, filename, config_file):
        # None of these would work:
        return JavaModuleLinter.create_arguments() + "my_module"
        return JavaModuleLinter._interface.create_arguments() + "my_module"
```
why this doesn't work?
the metaclass constructor is even called for derived classes
so when creating `MyJavaLinter`, we wrap the old already generated stuff inside `JavaModuleLinter` again!
this is extreme boilerplate we generate there
second
as the metaclass constructor is executed on derived classes, both versions inside `MyJavaLinter.create_arguments()` don't work
The first one: We wrapped the old interface into `JavaModuleLinter._interface`, `create_arguments` was moved into a different scope
The second one: Even we address the `_interface` now, it contains our current `create_arguments` from `MyJavaLinter`, so this would call recursively until your CPU dies^^
```
>>> class X(type):
...  def __init__(cls, name, bases, dct):
...   print("METACLASS")
...   type.__init__(cls, cls, name, bases)
...
>>> class Y(metaclass=X):
...  pass
...
METACLASS
>>> class Z(Y):
...  pass
...
METACLASS
```
using a decorator that takes care of generating everything is way better, as one sees that he generates the linter only on the class where the decorator is defined, and inheritance makes sense again.
```
@Linter(executable=...)
class MyLinter:
    def create_arguments(...):
        pass

class X(MyLinter):
    # no bad surprises, we know @Linter does some stuff with our class so we would look what explicitly
    pass
```
the only thing that makes stuff only aesthetically nicer with metaclasses, is `repr`
as `@Linter` returns a custom class, repr would yield.
```
>>> def a():
...  class X:
...   pass
...  return X
...
>>> a()
<class '__main__.a.<locals>.X'>
```
just not beautiful.
a metaclass could override `__repr__`, to print something better:
```
>>> class MetaLinter(type):
...  def __repr__(self):
...   return self.__name__
...
>>> class W(metaclass=MetaLinter):
...  pass
...
>>> W
W
```

in this case we could print something like
`<MyLinterBear executing 'java'>`

or similar
but the decorator would be needed anyway, so this would be just sugar, and can be filed inside a medium difficulty issue
alright that's about it, I fix tomorrow little stuff from my old `@Linter` decorator PR so it's ready again for review
okay, blasted the chat, looking into notifications (if any) and then I'm done
cya :3

**Comment by AbdealiJK** [comment]
Wow. It took me 15 mins to copy that with formatting ...
I've not read everything in super detail, but while copying it I found these things:

#### An example you gave seems wrong
You tried to make a class which has metaclass `type` inside the `__new__` - which is a weird way of doing things. I think the nicer way would be:
https://github.com/SublimeLinter/SublimeLinter3/blob/master/lint/linter.py#L66 or
https://github.com/SublimeLinter/SublimeLinter3/blob/master/lint/linter.py#L153
So, basically the functions are already created in `cls` and can be used. In fact you can even throw an error (when **defining** the class) if that function is not there. Which is super cool.

#### Inheritance
There is a very simply solution ... Just check `bases`.
https://github.com/SublimeLinter/SublimeLinter3/blob/master/lint/linter.py#L60

**Comment by Makman2** [comment]
I've thought about my idea again completely wrapping the user-given-interface into a member variable. I think inheritance is just simpler.

> You tried to make a class which has metaclass type inside the __new__ - which is a weird way of doing things.

I tried to wrap the stuff into a new member. `cls` is not the class of the user-defined-interface, it's the metaclass inside `__new__` (inside `__init__` it is indeed the already created class from the metaclass). Setting attributes on the metaclass doesn't make much sense. But inside `__new__` the given class is not yet created, and because I bypass creation by creating a different class, I need to define it still with `type(...)`.

I can still throw an error inside a decorator, I can just check `callable(userclass.create_arguments)`.

> There is a very simply solution ... Just check bases.

If I don't need control over the creation of subclasses, imo metaclasses are the wrong approach for this. A decorator is perfectly then, it works one-time on the given class.

Another advantage why I look the creation of a class inside a decorator is that I can hide some implementation details, like the compiled regex or so, imo that's somethine the user shouldn't have control. I know in python we are all adults, but I think this is a nice side-effect and prevents accidental misuse of such values.

**Comment by AbdealiJK** [comment]
Ok, so it seems Im a bit confused about metaclasses. But it does seem as thought metaclasses are a super set of decorators (i.e. it can do more) - http://stackoverflow.com/questions/1779372/python-metaclasses-vs-class-decorators

If you still think decorators are easier to go with - then let's continue with decorators - as I don't understand metaclasses well enough to suggest further improvements or alternatives :P

**Comment by Makman2** [comment]
Metaclasses give you control over the creation of a class, and also of it's bases. But in this case a decorator is easier imo :+1:

**Comment by AbdealiJK** [comment]
@Makman2 seems like a lot of issues depend on this as it revamps the whole Lint stuff

**Comment by Makman2** [comment]
Im continuing on thia tomorrow, thebn i've written my last exam :D

**Comment by AbdealiJK** [comment]
I've added comments on https://github.com/coala-analyzer/coala/commit/2173f8160526f1711605e45d84204505cc80b043 (Just mentioning it here so it does not get lost)
